Dockerfile

```diff
@@ -0,0 +1,22 @@
+# docker build -t contributions-service:latest .
+# docker run --init --rm -p 3030:3030 contributions-service:latest
+
+FROM node:12-alpine
+
+RUN apk add yarn
+
+WORKDIR /usr/src/app
+
+# Copy+install deps first to take advantage of layer caching
```

Comment:
Interesting - I want to learn more about this!

Reply:
My (limited) understanding of how it works is: each (non-empty) line in your Dockerfile is built as a 'layer', which is cached for subsequent builds. But if a cache is invalidated then all subsequent lines get rebuilt too. So you want commands that are more cacheable early on to avoid this as much as possible. Here, the expectation is that our npm packages will change less frequently than our `src` code.

At least, that's what I think is the answer! I copied some of this from an example node Dockerfile somewhere.

Reply:
My (limited) understanding of how it works is: each (non-empty) line in your Dockerfile is built as a 'layer', which is cached for subsequent builds. But if a cache is invalidated then all subsequent lines get rebuilt too. So you want commands that are more cacheable early on to avoid this as much as possible. Here, the expectation is that our npm packages will change less frequently than our `src` code.

At least, that's what I think is the answer! I copied some of this from an example node Dockerfile somewhere.

Reply:
Ah nice, yeah now that you mention it that sounds familiar!